Backpacks.cs

```diff
@@ -283,11 +329,26 @@ private void OnGroupPermissionRevoked(string group, string perm)
             {
                 foreach (IPlayer player in covalence.Players.Connected.Where(p => permission.UserHasGroup(p.Id, group)))
                 {
-                    DestroyGUI(player.Object as BasePlayer);
+                    if (!permission.UserHasPermission(player.Id, GUIPermission))
+                    {
+                        DestroyGUI(player.Object as BasePlayer);
+                    }
                 }
             }
         }
 
+        private void OnUserGroupAdded(string userId, string groupName)
+        {
+            if (permission.GroupHasPermission(groupName, GUIPermission))
+                CreateGUI(BasePlayer.Find(userId));
+        }
+
+        private void OnUserGroupRemoved(string userId, string groupName)
+        {
+            if (permission.GroupHasPermission(groupName, GUIPermission) && !permission.UserHasPermission(userId, GUIPermission))
+                DestroyGUI(BasePlayer.Find(userId));
+        }
+
```

Comment:
I'm hesitant to add more permission hooks. There are now six methods dealing with permissions being granted or revoked.

- `OnGroupPermissionGranted`
- `OnGroupPermissionRevoked`
- `OnUserGroupAdded`
- `OnUserGroupRemoved`
- `OnUserPermissionGranted`
- `OnUserPermissionRevoked`

I understand that using them all allows this plugin to achieve 100% accurate synchronization between GUI button visibility and permissions, but using this many hooks for simply a GUI button seems imbalanced in terms of effort/code compared to the benefit.

These sorts of hooks used to be in the plugin for resizing backpack capacity, but ultimately they were removed in favor of simply resizing when the backpack is opened, which really simplified things.

Can you explain more about why you think this is necessary to do? Are there example servers that dynamically assign and revoke the GUI button permission? If so, I'm curious to hear more about that use case since I wonder if there's a better way to solve the original problem. For example, if the use case is to grant/revoke the permission in specific zones or events, theoretically there's a more direct way to solve that problem that might be preferable.

Long term, ideally Oxide or uMod could have a more direct hook like `OnUserPermissionChanged(IPlayer player, string[] addedPerms, string[] removedPerms)` which would run whenever a player gained or lost permissions, after taking into account both user and group permissions. If something like that is ever implemented, I would be more inclined to add it then. I'm not strictly opposed to adding this now, but I would like to hear more information first.

Reply:
I don't really care much about those hooks. Added them just because there were already hooks that you mentioned earlier. If you prefer - you may just move GUI creation/deletion to backpack opening just like you did with resizing. The only thing is important from my point of view - it's destroying GUI on sleep/death and then create on respawn/sleep ends for those, who activated the button to avoid creation of GUI on dead a sleeping players screens.